File: codengine-ts/codengine-core-ts/src/task-definition.ts
// The description contract: the task-definition types an analyzer produces and a
// generator consumes. Mirror of codengine-spec/schema/task-definition.schema.json.

export type Kind = "number" | "boolean" | "string" | "array" | "object" | "any";

export interface Param {
  name: string;
  kind: Kind;
  required: boolean;
  nullable: boolean;
  default?: unknown;
}

export interface TaskDefinition {
  name: string;
  params: Param[];
  acceptsExtra: boolean;
}

export interface TaskDefinitions {
  version: "1";
  language: string;
  definitions: TaskDefinition[];
}

File: codengine-py/codengine-core-py/codengine_core/task_definition.py
"""The codengine description contract: the task-definition types an analyzer produces
and a generator consumes. Mirror of
codengine-spec/schema/task-definition.schema.json."""

from typing import Any, Literal, TypedDict

Kind = Literal["number", "boolean", "string", "array", "object", "any"]


class Param(TypedDict, total=False):
    """A named parameter. `default` is present only for an optional parameter."""

    name: str
    kind: Kind
    required: bool
    nullable: bool
    default: Any


class TaskDefinition(TypedDict):
    name: str
    params: list[Param]
    acceptsExtra: bool


class TaskDefinitions(TypedDict):
    version: str
    language: str
    definitions: list[TaskDefinition]
